File: Week_04/874/walking_robot_simulation.go
package main

type point struct {
	x, y int
}

func robotSim(commands []int, obstacles [][]int) int {
	dx, dy := 0, 1
	curr := point{0, 0}
	obstaclesMap := make(map[point]bool, 0)
	for i := 0; i < len(obstacles); i++ {
		obstaclesMap[point{obstacles[i][0], obstacles[i][1]}] = true
	}
	res := 0
	for i := 0; i < len(commands); i++ {
		if commands[i] == -2 {
			dx, dy = -dy, dx
		} else if commands[i] == -1 {
			dx, dy = dy, -dx
		} else {
			x, y := curr.x, curr.y
			for j := 0; j < commands[i]; j++ {
				x += dx
				y += dy
				if obstaclesMap[point{x, y}] {
					x -= dx
					y -= dy
					break
				}
			}
			curr.x, curr.y = x, y
			distance := curr.x*curr.x + curr.y*curr.y
			if distance > res {
				res = distance
			}
		}
	}
	return res
}

File: Week_04/874/walking_robot_simulation.py
from typing import List


class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        res = 0
        x, y, dx, dy = 0, 0, 0, 1
        obstacle_set = {
            (o[0], o[1]) for o in obstacles
        }
        for c in commands:
            if c == -2:
                dx, dy = -dy, dx
            elif c == -1:
                dx, dy = dy, -dx
            else:
                next_x, next_y = x, y
                for i in range(c):
                    next_x += dx
                    next_y += dy
                    if (next_x, next_y) in obstacle_set:
                        next_x -= dx
                        next_y -= dy
                        break
                x, y = next_x, next_y
                distance = x * x + y * y
                if distance > res:
                    res = distance
        return res
